### PR TITLE
Add support for converting CDF actions in DeltaSharingReader.

### PR DESCRIPTION
**Pull request by chakankardb**
This PR has the following changes:

1) Adds CDF actions (AddCdcFile, RemoveFile)
2) Convert CDF actions into pandas.
3) Unit tests.

It still does not have the rest client support for the query table changes RPC.

### CODE REVIEW

File: python/delta_sharing/reader.py [hunk 1]
```diff
@@ -93,6 +93,22 @@ def to_pandas(self) -> pd.DataFrame:
...
+        response = self._rest_client.list_table_changes(self._table, cdfOptions)
+
+        schema_json = loads(response.metadata.schema_string)
+
+        if len(response.actions) == 0 or self._limit == 0:
+            return get_empty_table(schema_json)
+
+        converters = to_converters(schema_json)
+        pdfs = []
+        for action in response.actions:
```
Comment (linzhou-db) [inline python/delta_sharing/reader.py:106]: Are we able to tell the class of these actions? For 'AddFile' and 'RemoveFile', we need to manually add column `_change_type`, as 'insert' or 'remove'.

File: python/delta_sharing/tests/test_reader.py [hunk 2]
```diff
@@ -237,3 +241,164 @@ def list_files_in_table(
...
+
+def test_table_changes_to_pandas_non_partitioned(tmp_path):
+    # Create basic data frame.
+    pdf1 = pd.DataFrame({"a": [1, 2, 3], "b": ["a", "b", "c"]})
+    pdf2 = pd.DataFrame({"a": [4, 5, 6], "b": ["d", "e", "f"]})
+    pdf3 = pd.DataFrame({"a": [7, 8, 9], "b": ["x", "y", "z"]})
+    pdf4 = pd.DataFrame({"a": [7, 8, 9], "b": ["x", "y", "z"]})
+
+    # Add change type (which is present in the parquet files).
+    pdf1["_change_type"] = "Insert"
```
Comment (linzhou-db) [inline python/delta_sharing/tests/test_reader.py:254]: For `insert/remove`, they are not from parquet files, and nit: first letter is not capital

File: python/delta_sharing/reader.py [hunk 3]
```diff
@@ -105,30 +121,48 @@ def _copy(
...
                     else:
                         raise ValueError("Cannot partition on binary or complex columns")
                 else:
                     pdf[col] = None
 
+        # If available, add timestamp and version columns from the action.
+        # All rows of the dataframe will get the same value.
+        if action.timestamp is not None:
+            assert DeltaSharingReader._current_timestamp_col_name() not in pdf.columns
+            pdf[DeltaSharingReader._current_timestamp_col_name()] = action.timestamp
```
Comment (linzhou-db) [inline python/delta_sharing/reader.py:154]: And maybe we should convert the `long` timestamp into string format.

File: python/delta_sharing/tests/test_rest_client.py [hunk 4]
```diff
@@ -378,3 +379,13 @@ def test_list_files_in_table_partitioned_different_schemas(
...
     ]
+
+
+def test_list_table_changes(
+    rest_client: DataSharingRestClient,
+):
+    try:
+        rest_client.list_table_changes(Table(name="x", share="y", schema="z"), CdfOptions())
+        assert False
+    except Exception as e:
```
Comment (linzhou-db) [inline python/delta_sharing/tests/test_rest_client.py:390]: This is to test the 'not implemented' error?

File: python/delta_sharing/reader.py [hunk 5]
```diff
@@ -93,6 +93,22 @@ def to_pandas(self) -> pd.DataFrame:
             copy=False,
         )[[field["name"] for field in schema_json["fields"]]]
 
+    def table_changes_to_pandas(self, cdfOptions: CdfOptions) -> pd.DataFrame:
+        response = self._rest_client.list_table_changes(self._table, cdfOptions)
+
+        schema_json = loads(response.metadata.schema_string)
+
+        if len(response.actions) == 0 or self._limit == 0:
```
[the remark this answers is not in the document]
Comment (chakankardb) [reply python/delta_sharing/reader.py]: Done.

File: python/delta_sharing/reader.py [hunk 6]
```diff
@@ -105,30 +121,48 @@ def _copy(
...
+            pdf[DeltaSharingReader._current_timestamp_col_name()] = action.timestamp
+
+        if action.version is not None:
+            assert DeltaSharingReader._current_version_col_name() not in pdf.columns
+            pdf[DeltaSharingReader._current_version_col_name()] = action.version
+
         return pdf
+
+    @staticmethod
+    def _current_timestamp_col_name():
```
[the remark this answers is not in the document]
Comment (chakankardb) [reply python/delta_sharing/reader.py:174]: Those are the column names the functions returns. The functions themselves is called *_col_name().

File: python/delta_sharing/reader.py (hunk 1 above)
Comment (chakankardb) [reply python/delta_sharing/reader.py:106]: Made changes to overwrite the change type as discussed.

File: python/delta_sharing/reader.py (hunk 3 above)
Comment (chakankardb) [reply python/delta_sharing/reader.py:158]: As discussed, keep it in the same format as input.
The client can always process the column and convert it as needed.

File: python/delta_sharing/tests/test_rest_client.py (hunk 4 above)
Comment (chakankardb) [reply python/delta_sharing/tests/test_rest_client.py:390]: Yes.
I'm implementing this in the next PR.

File: python/delta_sharing/reader.py (hunk 3 above)
Comment (linzhou-db) [reply python/delta_sharing/reader.py:158]: sg, let's revisit when necessary.

File: python/delta_sharing/reader.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (chakankardb) [reply python/delta_sharing/reader.py]: Done.

File: python/delta_sharing/reader.py [hunk 7]
```diff
@@ -105,30 +121,58 @@ def _copy(
...
                     if converter is not None:
-                        pdf[col] = converter(add_file.partition_values[col])
+                        pdf[col] = converter(action.partition_values[col])
                     else:
                         raise ValueError("Cannot partition on binary or complex columns")
                 else:
                     pdf[col] = None
 
+        # Add the change type col name to non cdc actions.
+        if type(action) != AddCdcFile:
```
[the remark this answers is not in the document]
Comment (chakankardb) [reply python/delta_sharing/reader.py]: Good catch! 
Done.

File: python/delta_sharing/reader.py [hunk 8]
```diff
@@ -93,6 +93,22 @@ def to_pandas(self) -> pd.DataFrame:
             copy=False,
         )[[field["name"] for field in schema_json["fields"]]]
 
+    def table_changes_to_pandas(self, cdfOptions: CdfOptions) -> pd.DataFrame:
+        response = self._rest_client.list_table_changes(self._table, cdfOptions)
+
+        schema_json = loads(response.metadata.schema_string)
+
+        if len(response.actions) == 0:
+            return get_empty_table(schema_json)
```
[the remark this answers is not in the document]
Comment (chakankardb) [reply python/delta_sharing/reader.py]: Good catch!
Done, and added some unit tests.

File: python/delta_sharing/reader.py [hunk 9]
```diff
@@ -105,30 +122,70 @@ def _copy(
 
     @staticmethod
     def _to_pandas(
-        add_file: AddFile, converters: Dict[str, Callable[[str], Any]], limit: Optional[int]
+        action: FileAction,
+        converters: Dict[str, Callable[[str], Any]],
+        for_cdc: bool,
```
Comment (linzhou-db) [inline python/delta_sharing/reader.py:127]: nit: `for_cdf` to keep consistency?

File: python/delta_sharing/reader.py [hunk 10]
```diff
@@ -105,30 +122,70 @@ def _copy(
...
-                        pdf[col] = converter(add_file.partition_values[col])
+                        pdf[col] = converter(action.partition_values[col])
                     else:
                         raise ValueError("Cannot partition on binary or complex columns")
                 else:
                     pdf[col] = None
 
+        if for_cdc:
+            # Add the change type col name to non cdc actions.
+            if type(action) != AddCdcFile:
```
Comment (linzhou-db) [inline python/delta_sharing/reader.py:156]: nit: we can use `if for_cdf and type(action) != AddCdcFile`?

**Review by linzhou-db — APPROVED** [review]

File: python/delta_sharing/reader.py (hunk 9 above)
Comment (chakankardb) [reply python/delta_sharing/reader.py:127]: Done.

File: python/delta_sharing/reader.py (hunk 10 above)
Comment (chakankardb) [reply python/delta_sharing/reader.py:156]: I'm skipping processing timestamp/version processing as well.

**Pull request merged**